Catalyst2 support (#5155)
ECP-WarpX/WarpX: Docs/source/dataanalysis/catalyst.rst

.. _visualization-catalyst:

In situ Visualization with Catalyst 2
=====================================
Catalyst 2 (further referred to as just Catalyst) is a lightweight in-situ visualization and analysis framework API developed for simulations and other scientific data producers. It has a lightweight implementation
(or **stub**) and an SDK to develop custom implementations of Catalyst. ParaView comes with its own implementation (known as **ParaView Catalyst**) for leveraging ParaView's
visualization and analysis capabilities, which is what this document will focus on.

Enabling Catalyst
-----------------

> In order to use Catalyst with WarpX, you must `build Catalyst 2 <https://catalyst-in-situ.readthedocs.io/en/latest/build_and_install.html>`_ and `build <https://github.com/Kitware/ParaView/blob/master/Documentation/dev/build.md>`_ or `install <https://www.paraview.org/download/>`_ ParaView 5.9+. Afterward, AMReX must be built with ``AMReX_CONDUIT=TRUE``,

Reviewer: Anonymous links for the general keywords (build, install) to avoid collisions in the sphinx reference of links:
```suggestion
In order to use Catalyst with WarpX, you must `build Catalyst 2 <https://catalyst-in-situ.readthedocs.io/en/latest/build_and_install.html>`_ and `build <https://github.com/Kitware/ParaView/blob/master/Documentation/dev/build.md>`__ or `install <https://www.paraview.org/download/>`__ ParaView 5.9+. Afterward, AMReX must be built with ``AMReX_CONDUIT=TRUE``,
```

``AMReX_CATALYST=TRUE``, ``Conduit_DIR=/path/to/conduit``, and ``Catalyst_DIR=/path/to/catalyst`` (``/path/to/catalyst`` should be the directory containing ``catalyst-config.cmake``, not the path to the implementation).

Once AMReX is appropriately built, WarpX can be built with the following options:

.. code-block:: cmake

>     WarpX_amrex_internal=FALSE
>     AMReX_DIR="/path/to/amrex/build"
> 
> If they cannot be found, ``Conduit_DIR`` and ``Catalyst_DIR`` will have to be set again. Ensure that AMReX is built with all required options, some common ones being:
> 
> .. code-block:: cmake
>     AMReX_MPI=TRUE

Reviewer: I think those blocks need an empty newline before the code starts
```suggestion

    WarpX_amrex_internal=FALSE
    AMReX_DIR="/path/to/amrex/build"

If they cannot be found, ``Conduit_DIR`` and ``Catalyst_DIR`` will have to be set again. Ensure that AMReX is built with all required options, some common ones being:

.. code-block:: cmake

    AMReX_MPI=TRUE
```

In addition to configuring the diagnostics, the following parameters must be included:
    * ``catalyst.script_paths``: The locations of the pipeline scripts, separated by either a colon or semicolon (e.g. ``/path/to/script1.py;/path/to/script2.py``).
    * ``catalyst.implementation`` (default ``paraview``): The name of the implementation being used (case sensitive).
    * ``catalyst.implementation_search_paths``: The locations to search for the given implementation. The specific file being searched for will be ``catalyst_{implementation}.so``.

Because the scripts and implementations are global, Catalyst does not benefit from nor differentiate between multiple diagnostics.

Visualization/Analysis Pipeline Configuration
---------------------------------------------
Catalyst uses the files specified in ``catalyst.script_paths`` to run all analysis.

The following script, :code:`simple_catalyst_pipeline.py`, automatically detects the type of data for both the mesh and particles, then creates an extractor for them. In most
cases, these will be saved as ``.VTPC`` files which can be read with the ``XML Partitioned Dataset Collection Reader``.

.. code-block:: python

>     from paraview.simple import *

Reviewer: ```suggestion
.. code-block:: python

    from paraview.simple import *
```

    def catalyst_finalize():
        return

    if __name__ == '__main__':
        paraview.simple.SaveExtractsUsingCatalystOptions(options)

For the case of ParaView Catalyst, pipelines are run with ParaView's included ``pvbatch`` executable and use the ``paraview`` library to modify the data. While pipeline scripts
could be written manually, this is not advised for anything beyond the script above. It is much more practical to use ParaView's built in ``Save Catalyst State`` button.

The process for creating a pipeline is as follows:
    1. Run at least one step of simulation and save the data in a ParaView compatible format, then open it in ParaView.
    2. Set up the desired scene, including filters, camera and views, and extractors.
    3. Press ``Save Catalyst State``, or the multicolored flask icon in the top left corner, and save it to a desired location.
    4. Open the script and replace the used producer with ``PVTrivialProducer``, setting the ``registrationName`` to either ``mesh`` or ``particles`` based on what data is used.

As an example for step four, here are a few lines from a script directly exported from ParaView:

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    # create a new 'XML Image Data Reader'
    meshdatavti = XMLImageDataReader(registrationName='meshdata.vti', FileName=['/path/to/meshdata.vti'])
    meshdatavti.CellArrayStatus = ['Bx', 'By', 'Bz', 'Ex', 'Ey', 'Ez']
    meshdatavti.TimeArray = 'None'

    # Calculator sample filter
    calculator1 = Calculator(registrationName='Calculator1', Input=meshdatavti)
    calculator1.AttributeType = 'Cell Data'
    calculator1.ResultArrayName = 'BField'
    calculator1.Function = 'sqrt(Bx^2 + By^2 + Bz^2)'

In order to use it with the mesh data coming from the simulation, the above code would be changed to:

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

.. _visualization-catalyst:

In situ Visualization with Catalyst 2
=====================================
Catalyst 2 (further referred to as just Catalyst) is a lightweight in-situ visualization and analysis framework API developed for simulations and other scientific data producers. It has a lightweight implementation
(or **stub**) and an SDK to develop custom implementations of Catalyst. ParaView comes with its own implementation (known as **ParaView Catalyst**) for leveraging ParaView's
visualization and analysis capabilities, which is what this document will focus on.

Enabling Catalyst
-----------------
In order to use Catalyst with WarpX, you must `build Catalyst 2 <https://catalyst-in-situ.readthedocs.io/en/latest/build_and_install.html>`_ and `build <https://github.com/Kitware/ParaView/blob/master/Documentation/dev/build.md>`__ or `install <https://www.paraview.org/download/>`__ ParaView 5.9+. Afterward, AMReX must be built with ``AMReX_CONDUIT=TRUE``,

> ``AMReX_CATALYST=TRUE``, ``Conduit_DIR=/path/to/conduit``, and ``Catalyst_DIR=/path/to/catalyst`` (``/path/to/catalyst`` should be the directory containing ``catalyst-config.cmake``, not the path to the implementation).

Reviewer: I am on the edge if the install details of how to set CMake hints, besides the options `AMReX_CONDUIT` and `AMReX_CATALYST`.

`/path/to/catalyst`: usually that is the path to the prefix/root of a package install path, if installed correctly:
https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure

`_DIR` variables: correct, but usually we just add the prefixes to the `CMAKE_PREFIX_PATH` environment variable.

I am ok to leave it for now.

Once AMReX is appropriately built, WarpX can be built with the following options:

.. code-block:: cmake

    WarpX_amrex_internal=FALSE

>     AMReX_DIR="/path/to/amrex/build"

Reviewer: Since we build AMReX as a superbuild by default, I would rather document to pass the `-DAMReX_CONDUIT=ON -DAMReX_CATALYST=ON` flags to a WarpX build.

(We do support external AMReX installs, of course, but do not do so outside of package managers.)

In addition to configuring the diagnostics, the following parameters must be included:
    * ``catalyst.script_paths``: The locations of the pipeline scripts, separated by either a colon or semicolon (e.g. ``/path/to/script1.py;/path/to/script2.py``).
    * ``catalyst.implementation`` (default ``paraview``): The name of the implementation being used (case sensitive).
    * ``catalyst.implementation_search_paths``: The locations to search for the given implementation. The specific file being searched for will be ``catalyst_{implementation}.so``.

Because the scripts and implementations are global, Catalyst does not benefit from nor differentiate between multiple diagnostics.

Visualization/Analysis Pipeline Configuration
---------------------------------------------
Catalyst uses the files specified in ``catalyst.script_paths`` to run all analysis.

The following script, :code:`simple_catalyst_pipeline.py`, automatically detects the type of data for both the mesh and particles, then creates an extractor for them. In most
cases, these will be saved as ``.VTPC`` files which can be read with the ``XML Partitioned Dataset Collection Reader``.

> .. code-block:: python

Reviewer: For such scripts, I would prefer writing them in an actual `.py` file and doing a `literalinclude`:
```rst
.. literalinclude:: catalyst/simple_catalyst_pipeline.py
   :language: python
   :caption: You can copy this file from ``Docs/source/dataanalysis/simple_catalyst_pipeline.py``.
```

That way, you can also use the documented scripts in our CI tests, to ensure they always work.

    from paraview.simple import *
    from paraview import catalyst

>     # Helper function

Reviewer: For such scripts, I would prefer writing them in an actual `.py` file and doing a `literalinclude`:
```
.. literalinclude:: catalyst/simple_catalyst_pipeline.py`
   :language: python
   :caption: You can copy this file from ``Docs/source/dataanalysis/simple_catalyst_pipeline.py``.
```

>     # Helper function
>     def create_extractor(data_node, filename="Dataset"):

Reviewer: For `create_extractor` I cannot follow from the doc string what it does yet - can add a more verbose string here?
Feel free to use the usual
```suggestion
    def create_extractor(data_node, filename="Dataset"):
        """A data extractor stores the pipeline data results instead of rendering a scene."""
```
syntax.
